src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/QueryGenerators/SqlQueryGenerator.cs

```diff
@@ -798,6 +803,74 @@ private void HandleTableKindSort(SearchParamTableExpression searchParamTableExpr
                 using (var delimited = StringBuilder.BeginDelimitedWhereClause())
                 {
                     AppendHistoryClause(delimited);
+                    AppendMinOrMax(delimited, context);
+
+                    if (searchParamTableExpression.Predicate != null)
+                    {
+                        delimited.BeginDelimitedElement();
+                        searchParamTableExpression.Predicate.AcceptVisitor(searchParamTableExpression.QueryGenerator, GetContext());
+                    }
+
+                    // if continuation token exists, add it to the query
+                    if (continuationToken != null)
+                    {
+                        var sortOperand = searchSort == SortOrder.Ascending ? ">" : "<";
+
+                        delimited.BeginDelimitedElement();
+                        StringBuilder.Append("((").Append(sortColumnName, null).Append($" = ").Append(Parameters.AddParameter(sortColumnName, sortValue));
+                        StringBuilder.Append(" AND ").Append(VLatest.Resource.ResourceSurrogateId, null).Append($" > ").Append(Parameters.AddParameter(VLatest.Resource.ResourceSurrogateId, continuationToken.ResourceSurrogateId)).Append(")");
+                        StringBuilder.Append(" OR ").Append(sortColumnName, null).Append($" {sortOperand} ").Append(Parameters.AddParameter(sortColumnName, sortValue)).AppendLine(")");
+                    }
+
+                    AppendIntersectionWithPredecessor(delimited, searchParamTableExpression);
+                }
+            }
+
+            _sortVisited = true;
+        }
+
+        private void HandleTableKindSortWithFilter(SearchParamTableExpression searchParamTableExpression, SearchOptions context)
```

Comment:
This methd is like 90% the same as HandleTableKindSort. Can you consolidate a bit?

Reply:
Agreed. I did think of consolidating things in this PR. But since I am going to change the code in HandleTableKindSort (to support querying for resources with "null values for sort parameter") my plan was to do the refactoring in the next PR.

---
In reply to: [613565040](https://github.com/microsoft/fhir-server/pull/1797#discussion_r613565040) [](ancestors = 613565040)

Reply:
ok